src/executor.rs

```diff
@@ -199,3 +133,96 @@ impl Executor {
         Ok(())
     }
 }
+
+async fn task_result(
+    task: &mut Option<JoinHandle<Result<Option<ExitStatus>>>>,
+) -> Result<Option<ExitStatus>> {
+    match task {
+        Some(handle) => handle.await.unwrap(),
+        None => match AlwaysPending.await {},
+    }
+}
+
+/// A future that will never resolve
+struct AlwaysPending;
+
+impl std::future::Future for AlwaysPending {
+    type Output = std::convert::Infallible;
+
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        std::task::Poll::Pending
+    }
+}
+
+/// Start the given task/command
+fn start_task(task: Task, command: &mut Command) -> Result<Child> {
+    command
+        .kill_on_drop(true)
+        .env("RUST_BACKTRACE", if task.backtrace { "1" } else { "0" })
+        .spawn()
+        .context("failed to launch command")
+}
+
+/// Send all lines in the process' output
+async fn execute_task(
+    mut child: Child,
+    with_stdout: bool,
+    line_sender: LineSender,
+) -> Result<Option<ExitStatus>> {
+    let stderr = child
+        .stderr
+        .take()
+        .ok_or_else(|| anyhow!("child missing stderr"))?;
+
+    let stderr_sender = line_sender.clone();
+    let stderr = stream_consumer(stderr, CommandStream::StdErr, stderr_sender);
+
+    let stdout = if with_stdout {
+        let stdout = child
+            .stdout
+            .take()
+            .ok_or_else(|| anyhow!("child missing stdout"))?;
+        let stdout_sender = line_sender.clone();
+        Some(stream_consumer(stdout, CommandStream::StdOut, stdout_sender))
+    } else {
+        None
+    };
+
+    // either we wait on both stdout and stderr concurrently, or just stderr.
+    if let Some(stdout) = stdout {
+        tokio::try_join!(stdout, stderr)?;
+    } else {
+        stderr.await?;
+    }
```

Comment:
this should be one if

Reply:
It could be, but I wanted to keep stream "creation" and awaiting separate for clarity.

Reply:
this doesn't work on that case you are adding complexity, less branch the better.